### ISSUE TITLE
Allow custom JwtBearerEvents to allow SignalR to work with Bearer tokens.

### ISSUE THREAD

**Issue by jraadt** [okta/okta-aspnet#71]
OktaWebApiOptions should include an option to set Events (JwtBearerEvents). This is needed to allow SignalR to work with Bearer tokens. See the example at https://docs.microsoft.com/en-us/aspnet/core/signalr/authn-and-authz?view=aspnetcore-2.2

**Comment by nbarbettini** [comment]
Hey @jraadt! Let me know if I am understanding your scenario correctly:

1. Your client/frontend app signs a user in and gets a token from Okta
2. This token is used to make authorized API calls from the frontend to your API
3. This token is also used to establish a connection from the client to a SignalR hub on your API
4. The SignalR JS client doesn't send the `Authorization: Bearer` header (it sends tokens in the query string instead) so in this case you need to customize how the middleware pulls the token out of the request.

**Comment by jraadt** [comment]
@nbarbettini That's exactly right.

**Comment by chillitom** [comment]
👍 this would be extremely useful

is there a workaround?
will the associated PR be merged?

**Comment by chillitom** [comment]
Here's a workaround for anyone waiting on the PR.

```csharp
using Microsoft.AspNetCore.Authentication;
using Microsoft.AspNetCore.Authentication.JwtBearer;
using Microsoft.Extensions.DependencyInjection;
using Okta.AspNet.Abstractions;
using System;

namespace SignalR.Okta.Fix {

    // these customizations are required to get Okta and SignalR working nicely together
    // https://github.com/okta/okta-aspnet/issues/71
    // https://docs.microsoft.com/it-it/aspnet/core/signalr/authn-and-authz?view=aspnetcore-5.0

    public class ExtendedOktaWebApiOptions : OktaWebApiOptions {
        public JwtBearerEvents Events { get; set; }
    }

    public static class OktaAuthenticationOptionsExtensions {

        /// <summary>
        /// Configures Okta for Web API apps.
        /// </summary>
        /// <param name="builder">The application builder.</param>
        /// <param name="options">The Okta Web API options.</param>
        /// <returns>The authentication builder.</returns>
        public static AuthenticationBuilder AddOktaWebApi(this AuthenticationBuilder builder, ExtendedOktaWebApiOptions options) {
            if (builder == null) {
                throw new ArgumentNullException(nameof(builder));
            }

            new OktaWebApiOptionsValidator().Validate(options);

            return AddJwtValidation(builder, options);
        }

        private static AuthenticationBuilder AddJwtValidation(AuthenticationBuilder builder, ExtendedOktaWebApiOptions options) {
            var issuer = UrlHelper.CreateIssuerUrl(options.OktaDomain, options.AuthorizationServerId);

            var tokenValidationParameters = new DefaultTokenValidationParameters(options, issuer) {
                ValidAudience = options.Audience,
            };

            builder.AddJwtBearer(opt => {
                opt.Audience = options.Audience;
                opt.Authority = issuer;
                opt.TokenValidationParameters = tokenValidationParameters;
                opt.BackchannelHttpHandler = new OktaHttpMessageHandler("okta-aspnetcore", typeof(OktaAuthenticationOptionsExtensions).Assembly.GetName().Version, options);

                opt.SecurityTokenValidators.Clear();
                opt.SecurityTokenValidators.Add(new StrictSecurityTokenValidator());

                opt.Events = options.Events;
            });

            return builder;
        }
    }
}
```

**Comment by laura-rodriguez** [comment]
Hi @chillitom,

Thanks for your question and the workaround for folks that are waiting for this 💯 . 

We have a plan to expose all the events, but I can't share an ETA at the moment, since the .NET team is currently working on different projects. 

Thank you so much for your patience!

**Comment by laura-rodriguez** [comment]
OIDC and JWT events are now exposed. This feature is available in Okta.AspNet 2.0.0 and Okta.AspNetCore 4.0.0. Fixed in #187.

**Comment by thepherm** [comment]
Thank you!
